webapp/tests/karma/ts/services/telemetry.service.spec.ts

```diff
@@ -0,0 +1,265 @@
+      Object.defineProperty(window.screen, 'availHeight', { value: 1024 });
+
+      await service.record('test', 1);
+
+      expect(consoleErrorSpy.callCount).to.equal(0);
+      expect(pouchDb.post.callCount).to.equal(1);
+      expect(pouchDb.post.args[0][0]).to.deep.include({ key: 'test', value: 1 });
+      expect(dbInstance.put.callCount).to.equal(1);
+
+      const aggregatedDoc = dbInstance.put.args[0][0];
+      expect(aggregatedDoc._id).to.match(/telemetry-2018-10-greg/);
+      expect(aggregatedDoc.metrics).to.deep.equal({
+        foo: 'stats',
+        bar: 'more stats',
+      });
+      expect(aggregatedDoc.type).to.equal('telemetry');
+      expect(aggregatedDoc.metadata.year).to.equal(2018);
+      expect(aggregatedDoc.metadata.month).to.equal(10);
+      expect(aggregatedDoc.metadata.user).to.equal('greg');
+      expect(aggregatedDoc.metadata.versions).to.deep.equal({
+        app: '3.0.0',
+        forms: {
+          'anc_followup': '1-abc'
+        }
+      });
+      expect(aggregatedDoc.dbInfo).to.deep.equal({ some: 'stats' });
+      expect(aggregatedDoc.device).to.deep.equal({
+        userAgent: 'Agent Smith',
+        hardwareConcurrency: 4,
+        screen: {
+          width: 768,
+          height: 1024,
+        },
+        deviceInfo: {}
+      });
+      expect(dbInstance.query.callCount).to.equal(1);
+      expect(dbInstance.query.args[0][0]).to.equal('medic-client/doc_by_type');
+      expect(dbInstance.query.args[0][1]).to.deep.equal({ key: ['form'], include_docs: true });
+      expect(pouchDb.destroy.callCount).to.equal(1);
+      expect(pouchDb.close.callCount).to.equal(0);
+    });
+
+    it('should deal with conflicts by making the ID unique and noting the conflict in the new document', async () => {
+      storageGetItemStub.withArgs('medic-greg-telemetry-db').returns('dbname');
+      storageGetItemStub.withArgs('medic-greg-telemetry-date').returns(
+        moment()
+          .subtract(5, 'weeks')
+          .valueOf()
+          .toString()
+      );
+
+      pouchDb.post = sinon.stub().resolves();
+      pouchDb.query = sinon.stub().resolves({
+        rows: [
+          {
+            key: 'foo',
+            value: 'stats',
+          },
+          {
+            key: 'bar',
+            value: 'more stats',
+          },
+        ],
+      });
+      dbInstance.info.resolves({ some: 'stats' });
+      dbInstance.put.onFirstCall().rejects({ status: 409 });
+      dbInstance.put.onSecondCall().resolves();
+      dbInstance.get.withArgs('_design/medic-client').resolves({
+        _id: '_design/medic-client',
+        deploy_info: {
+          version: '3.0.0'
+        }
+      });
+      dbInstance.query.resolves({ rows: [] });
+      pouchDb.destroy = sinon.stub().callsFake(() => {
+        pouchDb._destroyed = true;
+        return Promise.resolve();
+      });
+      pouchDb.close = sinon.stub();
+
+      Object.defineProperty(window.navigator, 'userAgent', { value: 'Agent Smith' });
```

Comment:
Please, make sure to reset these when the test is done!

Reply:
Done!